### DISCUSSION TITLE
How to filter messages

### DISCUSSION THREAD

**Discussion by evarrica** [Havret/dotnet-activemq-artemis-client#422] [Q&A] [answered]
Good morning
In the Artemis MQ documentation you can filter the message by consumer thought filter expression [Address filter Artemis](https://activemq.apache.org/components/artemis/documentation/latest/address-model.html) where the java code is:

`MessageConsumer redConsumer = redSession.createConsumer(queue, "color='red'");`

What is the corresponding .net code?

**Comment by Havret** [comment, marked as answer]
Hi @evarrica, 

To filter messages you need to specify the `FilterExpression`:

```csharp
await using IConnection connection = await CreateConnection();
var address = Guid.NewGuid().ToString();

await using IConsumer redMessageConsumer = await connection.CreateConsumerAsync(new ConsumerConfiguration
{
    Address = address,
    RoutingType = RoutingType.Anycast,
    FilterExpression = "color = 'red'"
});
await using IConsumer blueMessageConsumer = await connection.CreateConsumerAsync(new ConsumerConfiguration
{
    Address = address,
    RoutingType = RoutingType.Anycast,
    FilterExpression = "color = 'blue'"
});
```

You can find more examples [here](https://github.com/Havret/dotnet-activemq-artemis-client/blob/821ee4ec5395ae07db4595751fc9d1538e13cc8d/test/ArtemisNetClient.IntegrationTests/FilterExpressionsSpec.cs).

Best,
Havret
